### PR TITLE
[data] Compute valid-token counts during collation

### PR DESCRIPTION
**Pull request by SherlockNoMad**
## Summary

`Trainer.train_step` scanned every microbatch's labels for `IGNORE_INDEX` on the host, before the batch was moved to the device. That put a full CPU pass over the labels of every microbatch on the training loop's critical path, once per optimizer step.

Collators already materialize the labels while building the batch, and they run in the Grain prefetch thread, so the count is effectively free there. `TextCollator` and `MultiModalCollator` now set `num_valid_tokens` as an ordinary field of the batch dict, which keeps the existing two-value batch interface intact.

The consumers pop the field before the batch reaches the model: `Trainer.train_step`, `Validator`, and the forge and torchft experiment trainers. Popping matters because `decoder.preprocess_inputs` pops only `input`/`labels` and forwards every remaining key as `**extra_kwargs` to `model()` — so leaving the field in place would send it into `forward()`, and into CP sharding and SPMD annotation on the way. With the pop, the dict the model receives is exactly what it expects.

Two knock-on cleanups: the validator now uses the popped count instead of its own device-side label scan and builds its reduction tensor once like the trainer, and `Trainer.train_step` is annotated with the `TrainerBatch` alias, since `dict[str, torch.Tensor]` no longer describes a dict holding an int.

`FluxCollator` does not set the field. Flux labels are raw image tensors rather than token targets, and Flux has its own trainer and validator that compute their own counts, so it is unaffected.

## Behavior change

The field is required, not optional. A batch reaching the core trainer or validator without it raises `KeyError` rather than silently rescanning the labels. That is a breaking change for out-of-tree dataloaders, but it fails loudly.

Popping mutates the caller's dict, which is consistent with the existing in-place `input_dict[key] = value.to(device)`. Grain yields a fresh dict per batch, so this is safe here; a dataloader that recycles batch dicts would break on the second read.

## Numerics

Unchanged. The count is the same integer, reduced the same way. The only difference is that `local_valid_tokens` accumulates as a Python `int` and is materialized once as a device tensor for the data-parallel all-reduce, instead of accumulating in a host tensor that was then copied to the device.

## Test plan

```
pytest -q tests/unit_tests/cpu --ignore=tests/unit_tests/cpu/test_torch_checkpointing.py
```
773 passed, 1 skipped. The single failure, `test_lora.py::test_lora_forward`, is a pre-existing Inductor flex-CPU error that reproduces on a clean tree.

New coverage: `TextCollator` and `MultiModalCollator` populate the field, and batches out of `GrainDataLoader` carry it end to end. Trainer, validator, and invalid-loss fixtures were updated to supply it — including three that reused a single dict object across microbatches, which the pop exposed.

`pre-commit run --files <changed files>` is clean.

## Review notes

Second commit addresses the feedback from @fegin, @jinsooihm and @tianyu-l: the `BatchInputsWithMetadata` dict subclass and the `get_batch_num_valid_tokens` helper are gone, replaced by the plain default field.

### CODE REVIEW

File: torchtitan/components/data/collators.py [hunk 1]
```diff
@@ -22,6 +22,52 @@
...
+def get_batch_num_valid_tokens(
+    input_dict: dict[str, Any],
+    labels: torch.Tensor,
+    *,
+    ignore_index: int,
+) -> int:
+    """Use a pipeline-provided token count, falling back to a local label scan."""
+    num_valid_tokens = getattr(input_dict, "num_valid_tokens", None)
+    if num_valid_tokens is not None:
+        return num_valid_tokens
```
Comment (fegin) [inline torchtitan/components/data/collators.py]: Will the `num_valid_tokens` get passed into the model forward()? If yes, this may cause issues for some models. If no errors will occurs, it is better to keep the `input_dict` as what the models expect.

**Review by fegin — COMMENTED** [review]
cc., @jinsooihm to review too

File: torchtitan/components/data/collators.py (hunk 1 above)
Comment (jinsooihm) [reply torchtitan/components/data/collators.py]: Also, it feels a little awkward that `input_dict` contains `num_valid_tokens` which is an information about the labels, but does not actually contain the labels.

Comment (tianyu-l) [reply torchtitan/components/data/collators.py]: I think we should put labels in `input_dict`, why not?

File: torchtitan/trainer.py [hunk 2]
```diff
@@ -834,26 +835,35 @@ def train_step(
...
         microbatch_groups: list[list[tuple[dict[str, torch.Tensor], torch.Tensor]]] = []
-        local_valid_tokens = torch.tensor(0, dtype=torch.int64)
+        local_valid_tokens = 0
         for _ in range(self.gradient_accumulation_steps):
             microbatches = []
             for _ in range(self.num_pp_microbatches):
                 with sl.log_trace_span("fetching_batch"):
                     input_dict, labels = next(data_iterator)
-                local_valid_tokens += (labels != IGNORE_INDEX).sum()
+                local_valid_tokens += get_batch_num_valid_tokens(
```
Comment (tianyu-l) [inline torchtitan/trainer.py]: Comment 1: compared with having a standalone util function + dataclass, I'd rather just making it a default field for (text) dataloader.

Comment 2: Instead of loading all microbatches -> compute global_valid_tokens -> start training, I think @felipemello1 wants the capability of "train-microbatch-as-you-go" without knowing exact loss scale, and later adjust the gradient. I didn't embrace the idea, but maybe it'd become absolute necessary. Anyway that's another direction.

**Review by tianyu-l — COMMENTED** [review]
> 23 ms in the PP2 case for dsv3-671B with 32 microbatches.

How much would it give us, from 92.29%?

File: torchtitan/trainer.py (hunk 2 above)
Comment (SherlockNoMad) [reply torchtitan/trainer.py]: re 1. Agree that making this a default field is nicer . 

re 2. Current approach is divide the gradient before accumulation. The other approach is divide after accumulation. I think both works, I don't see the clear benefit of second approach tho.

**Comment by SherlockNoMad** [comment]
> > 23 ms in the PP2 case for dsv3-671B with 32 microbatches.
> 
> How much would it give us, from 92.29%?

Step time is ~10s, so this is ~0.2% for this case.

File: torchtitan/trainer.py (hunk 2 above)
Comment (felipemello1) [reply torchtitan/trainer.py]: Second approach allows overlapping. Today we do:

```python
[load data 1] [load data 2] [load data 3]
										     [train batch 1] [train batch 2] [train batch 3]
```

Instead, we could do:
```python
[load data 1] [load data 2] [load data 3]
			[train batch 1] [train batch 2] [train batch 3]
```

For pretraining, its not a big deal. We prefetch the data. Perhaps for multimodal it makes a difference?

But, for RL, this is a **major** bottleneck.

Here, "load data N" is actually **generating** it on the fly. So its very slow. In some settings, we can easily do >100 grad acc steps per train step. Waiting for all 100 batches being ready so we can start training means the trainer stays idle for a long time unnecessarily. 

The "prefetch" equivalent is higher off policy, which is undesirable in RL.

Let me know if that makes sense

Comment (SherlockNoMad) [reply torchtitan/trainer.py]: per offline discussion, I agree overlapping is needed for RL training. 
Let's do that in another PR? This PR would be just moving the valid token counting to dataloader.

Comment (fegin) [reply torchtitan/trainer.py]: Agree that overlapping deserves another PR.

**Review by fegin — APPROVED** [review]
